Refactor Test&Troubleshoot (#2666)
astronomer/docs: astro/cli/run-airflow-locally.md

---
sidebar_label: 'Run your project locally'
title: 'Run your Astro project in a local Airflow environment'
id: run-airflow-locally
description: Run commands in your local Airflow environment to troubleshoot running DAGs and tasks.
---

> Running your Astro project locally is an easy way to test your code changes. Additionally, you can only troubleshoot some issues in a project after running your local Airflow environment with `astro dev start`.

Reviewer: ```suggestion
Running Airflow locally with the Astro CLI can be an easy way to test DAG changes quickly before deploying your code to a Deployment on Astro. Testing a DAG locally first requires that you start a local Airflow environment on your computer by running the `astro dev start` command in your terminal.
```

I'm not super sure what the purpose of this 2nd sentence is? Is it supposed to be a prerequisite?

> Running your Astro project locally is an easy way to test your code changes. Additionally, you can only troubleshoot some issues in a project after running your local Airflow environment with `astro dev start`.

Reviewer 2: hopping in with some suggestions based on our standup conversation.
```suggestion
Running Airflow locally with the Astro CLI can be an easy way to preview and debug DAG changes quickly before deploying your code to a Deployment on Astro. By locally running your DAGs, you can iteratively develop and improve your DAGs in an Airflow without consuming infrastructure resources or waiting on code deploy processes.
```

> Running your Astro project locally is an easy way to test your code changes. Additionally, you can only troubleshoot some issues in a project after running your local Airflow environment with `astro dev start`.
> 
> This document explains how to use Astro CLI commands to interact with a locally running Airflow environment.

Reviewer 2: This might also be a place to link to Develop a Project, but I'm having a terrible time coming up with a one sentence summary of what that page contains, even with this refactoring.

```suggestion
This document explains how to use the Astro CLI to start a local Airflow environment on your computer and interact with your Astro project. To learn more about unit testing for your DAGs or testing project dependencies when changing Python or Astro Runtime versions, see [Test your project locally](test-your-astro-project-locally.md).
```

> This document explains how to use Astro CLI commands to interact with a locally running Airflow environment.

Author: ```suggestion
```

## Start a local Airflow environment

To begin running your project in a local Airflow environment, run:

```bash
astro dev start
```

> This command builds your project and spins up 4 containers on your machine, each for a different Airflow component. After the command completes, you can access your project's Airflow UI at `https://localhost:8080/`.

Reviewer: ```suggestion
This command builds your project and spins up 4 containers on your machine, each for a different Airflow component. After the command completes, you can access a local version of the Airflow UI at `https://localhost:8080/`. Any DAGs in the `dags` folder of your Astro project will appear.
```

Author: The last sentence is true, but what value does it add? Who is looking for that information, or would assume otherwise?

## Restart a local Airflow environment 

> To restart your local Airflow environment, run:

Reviewer: Would add 1 sentence about why you'd want to restart your environment

> To restart your local Airflow environment, run:

Author: ```suggestion
Restarting your Airflow environment rebuilds your image and restarts the Docker containers running on your local machine with the new image. Restart your environment to apply changes from specific files in your project, or to troubleshoot issues that occur when your project is running. 

To restart your local Airflow environment, run:
```

```sh
astro dev restart
```

This command rebuilds your image and restarts the Docker containers running on your local machine with the new image. Alternatively, you can run `astro dev stop` to stop your Docker containers without restarting your environment, then run `astro dev start` when you want to restart.

## Stop a local Airflow environment 

> Run the following to pause all Docker containers running your local Airflow environment. 

Reviewer: Same here. "Why" do users stop their Airflow environment?

> Run the following to pause all Docker containers running your local Airflow environment. 

Author: ```suggestion
Run the following command to pause all Docker containers running your local Airflow environment. 
```

```sh
astro dev stop
```

> Unlike `astro dev kill`, this command does not prune mounted volumes and delete data associated with your local Postgres database. If you run this command, Airflow connections and task history will be preserved.

Reviewer: Might have this mentioned before the command is listed, and when we explain why a user would use this

Reviewer: Also, let's link out the `astro dev kill` command to the section in this doc about it? Or make sure the language ("hard reset...") matches

Author: You use it to stop your environment. With the arrival of `astro dev restart`, it doesn't really have a purpose for local development besides that.. you want to stop.

> Unlike `astro dev kill`, this command does not prune mounted volumes and delete data associated with your local Postgres database. If you run this command, Airflow connections and task history will be preserved.

Author: ```suggestion
Unlike [`astro dev kill`](#hard-reset-your-airflow-environment), this command does not prune mounted volumes and delete data associated with your local Postgres database. If you run this command, Airflow connections and task history will be preserved.
```

> This command can be used regularly with `astro dev start` and `astro dev restart` during local development.

Reviewer: I see this sentence, but just not enough guidance

> This command can be used regularly with `astro dev start` and `astro dev restart` during local development.

Author: ```suggestion
```

## View Airflow component logs

> You can use the Astro CLI to view logs for your local Airflow environment's webserver, scheduler, and triggerer. This is useful if you want to troubleshoot a specific task instance, or if your environment suddenly stops working after a code change.

Reviewer: ```suggestion
You can use the Astro CLI to view logs for your local Airflow environment's webserver, scheduler, and triggerer. This is useful if you want to troubleshoot a specific task instance, or if your local environment does not run properly after a code change. For example, if the Airflow UI is not available or no scheduled tasks are running.
```

To view component logs in a local Airflow environment, run:

```sh

> astro dev logs

Reviewer: Should we add another example command to see logs for a specific component? Seems a more likely use case, and I think it's ok to take a bit more real-estate

```

See the [Astro CLI reference guide](cli/astro-dev-logs.md) for more details and options.

> ## Run Airflow CLI commands

Reviewer: Again -- what are the top 2-3 reasons people run Airflow CLI commands locally? Let's list them. @manmeetkaur's 2 cents might be helpful here

Author: I think the example after the command is sufficient! To be honest I don't love the idea of adding "why" to every single topic in this guide. Listing examples of commands people might use might frustrate someone who just needs to know how to run the specific Airflow command they currently need, which is how I assume most people are interacting with a doc like this.

---

> sidebar_label: 'Run your project locally'

Author: ```suggestion
sidebar_label: 'Run Airflow locally'
```

    apt-get install gcc
    ```

5. Open your Astro project `packages.txt` file and add the package references you removed in Step 1 individually until you find the package that is the source of the error.

> ## Override the Astro CLI Docker Compose file

Reviewer: Check formatting here?

Author: It's just GitHub failing to parse

- A failed Airflow or Astro Runtime version upgrade.
- Misconfigured Dockerfile or Docker override file.
- Misconfigured Airflow settings including `packages.txt` or `requirements.txt`.

### Ports are not available for my local Airflow webserver 

By default, the Astro CLI uses port `8080` for the Airflow webserver and port `5432` for the Airflow metadata database in a local Airflow environment. If these ports are already in use on your local computer, an error message similar to the following appears:

```text
Error: error building, (re)creating or starting project containers: Error response from daemon: Ports are not available: exposing port TCP 0.0.0.0:5432 → 0.0.0.0:0: listen tcp 0.0.0.0:5432: bind: address already in use
```

To resolve a port availability error, you have the following options:

- Stop all running Docker containers and restart your local environment using `astro dev restart`.
- Change the default ports for these components. For example, you can use `astro config set webserver.port 8081` for the webserver and `astro config set postgres.port 5433` for Postgres. See [Configure CLI](cli/configure-cli.md) for all available configurations.

> ### Stop all running Docker containers

Author: ####

Make sure to specify `version: "3.1"` and follow the format of the source code file linked above.

Run the following command to see the override file in your scheduler container:

```sh
astro dev bash --scheduler "ls -al"
```

> ## Common issues 

Reviewer: Let's add a sentence here introducing this section? We can link out to support or our forum too

Make sure to specify `version: "3.1"` and follow the format of the source code file linked above.

Run the following command to see the override file in your scheduler container:

```sh
astro dev bash --scheduler "ls -al"
```

## Common issues 

> 

Author: ```suggestion
Use the following topics to resolve common issues with running an Astro project in a local environment.
```

> Use the following topics to resolve common issues with running an Astro project in a local environment.

Author: ```suggestion

Use the following topics to resolve common issues with running an Astro project in a local environment.

```

To resolve a port availability error, you have the following options:

- Stop all running Docker containers and restart your local environment using `astro dev restart`.
- Change the default ports for these components. For example, you can use `astro config set webserver.port 8081` for the webserver and `astro config set postgres.port 5433` for Postgres. See [Configure CLI](cli/configure-cli.md) for all available configurations.

> ### Stop all running Docker containers

Author: ```suggestion
#### Stop all running Docker containers
```

    - Run `docker stop <container_id>` to stop a specific Docker container. Replace `<container_id>` with one of the values you copied in step 2.
    - Run `docker stop $(docker ps -q)` to stop all running Docker containers.

> ### Change the default port assignment

Author: ```suggestion
#### Change the default port assignment
```
